test/e2e/fishapp/shared/dynamic_stack.go

```diff
@@ -0,0 +1,631 @@
+package shared
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	appmeshv1beta1 "github.com/aws/aws-app-mesh-controller-for-k8s/pkg/apis/appmesh/v1beta1"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/collection"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/k8s"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"io/ioutil"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+const (
+	connectivityCheckRate = time.Second / 100
+	appMeshOperationRate  = time.Second * 1
+)
+
+// An dynamic generated stack designed to test app mesh integration :D
+// Suppose we have configuration:
+//		5 VirtualServicesCount
+//		10 VirtualNodesCount
+//		2 RoutesCountPerVirtualRouter
+//		2 TargetsCountPerRoute
+//		4 BackendsCountPerVirtualNode
+// we will generate below virtual service configuration & virtual node configuration
+// =======virtual services =========
+//	vs1 -> /path1 -> vn1(50)
+//				  -> vn2(50)
+//		-> /path2 -> vn3(50)
+//				  -> vn4(50)
+//	vs2 -> /path1 -> vn5(50)
+//				  -> vn6(50)
+//		-> /path2 -> vn7(50)
+//				  -> vn8(50)
+//	vs3 -> /path1 -> vn9(50)
+//				  -> vn10(50)
+//		-> /path2 -> vn1(50)
+//				  -> vn2(50)
+//	vs4 -> /path1 -> vn3(50)
+//				  -> vn4(50)
+//		-> /path2 -> vn5(50)
+//				  -> vn6(50)
+//	vs5 -> /path1 -> vn7(50)
+//				  -> vn8(50)
+//		-> /path2 -> vn9(50)
+//				  -> vn10(50)
+// =======virtual nodes =========
+//  vn1 -> vs1,vs2,vs3,vs4
+//  vn2 -> vs5,vs1,vs2,vs3
+//  vn3 -> vs4,vs5,vs1,vs2
+//  ...
+//
+// then we validate each virtual node can access each virtual service at every path, and calculates the target distribution
+type DynamicStack struct {
+	// service discovery type
+	ServiceDiscoveryType ServiceDiscoveryType
+
+	// number of virtual service
+	VirtualServicesCount int
+
+	// number of virtual nodes count
+	VirtualNodesCount int
+
+	// number of routes per virtual router
+	RoutesCountPerVirtualRouter int
+
+	// number of targets per route
+	TargetsCountPerRoute int
+
+	// number of backends per virtual node
+	BackendsCountPerVirtualNode int
+
+	// number of replicas per virtual node
+	ReplicasPerVirtualNode int32
+
+	// how many time to check connectivity per URL
+	ConnectivityCheckPerURL int
+
+	// ====== runtime variables ======
+	mesh      *appmeshv1beta1.Mesh
+	namespace *corev1.Namespace
+
+	createdNodeVNs  []*appmeshv1beta1.VirtualNode
```

Comment:
will it make it easier to track these as maps instead of slices?

Reply:
maps works too, but slice is enough since all i need is `index -> node` :D